crates/ol/stf/src/verification.rs

```diff
@@ -0,0 +1,510 @@
+    if block_context.is_probably_epoch_initial() {
+        let epoch_context = block_context.to_epoch_initial_context();
+        chain_processing::process_epoch_initial(state, &epoch_context)?;
+    }
+
+    // 2. Process the slot start for every block.
+    chain_processing::process_slot_start(state, &block_context)?;
+
+    // 3. Call process_block_tx_segment for every block as usual.
+    let output_buffer = ExecOutputBuffer::new_empty();
+    let basic_ctx = BasicExecContext::new(block_info, &output_buffer);
+    let tx_ctx = TxExecContext::new(&basic_ctx, parent_header.as_ref());
+    transaction_processing::process_block_tx_segment(state, body.tx_segment(), &tx_ctx)?;
+
+    // 4. Check the state root.
+    // - if it's a nonterminal, then check against the header state root
+    // - if it *is* a terminal, then check against the preseal state root
+    let pre_manifest_state_root = state.compute_state_root()?;
+
+    if body.is_probably_terminal() {
+        // For terminal blocks, check against the preseal state root
+        let expected_preseal = exp
+            .post_state_roots
+            .preseal_state_root()
+            .ok_or(ExecError::ChainIntegrity)?;
+        if &pre_manifest_state_root != expected_preseal {
+            return Err(ExecError::ChainIntegrity);
+        }
+    } else {
+        // For non-terminal blocks, check against the header state root
+        if &pre_manifest_state_root != exp.post_state_roots.header_state_root() {
+            return Err(ExecError::ChainIntegrity);
+        }
+    }
+
+    // 5. If it's the last block of an epoch, then call process_block_manifests,
+    // then really check the header state root.
+    //
+    // Then we get the exec output one way or another.
+    if let Some(manifest_container) = body.l1_update().map(|u| u.manifest_cont()) {
+        manifest_processing::process_block_manifests(
+            state,
+            manifest_container,
+            tx_ctx.basic_context(),
+        )?;
+
+        // After processing manifests, check the actual final state root against the header.
+        let final_state_root = state.compute_state_root()?;
+        if &final_state_root != exp.post_state_roots.header_state_root() {
+            return Err(ExecError::ChainIntegrity);
+        }
+    }
+
+    // 6. Check the logs root.
+    let computed_logs_root = compute_logs_root(&output_buffer.into_logs());
+    if computed_logs_root != exp.logs_root {
+        return Err(ExecError::ChainIntegrity);
+    }
+
+    Ok(())
+}
+
+/// Checks that headers are properly continuous.
+pub fn verify_header_continuity(
+    header: &OLBlockHeader,
+    parent: Option<&OLBlockHeader>,
+) -> ExecResult<()> {
+    // Check parent linkages.
+    if let Some(ph) = parent {
+        // Simply check that the parent block makes sense.
+        let pblkid = ph.compute_blkid();
+        if *header.parent_blkid() != pblkid {
+            return Err(ExecError::BlockParentMismatch);
+        }
```

Comment:
Wait, how do we even get here? We get to this function from `verify_block_classically` which is passed a block (header and body) and a parent header -- what is the parent header? Isn't it the header of _the block (id) referenced in the block's header_? We already got this header from somewhere, say our db, so we already checked that the header and the block id makes sense.

If the purpose is merely a sanity check, we can have it, but let's make sure to write it is for sanity.